Fix unfinished multi-threaded batch analysis and memory leak

uklotzde opened the pull request:
[Conversation on Zulip](https://mixxx.zulipchat.com/#narrow/stream/109171-development/subject/analysis.20performance/near/154270414)

Unfortunately I was not able to trigger the error once again and I'm not able to reason about how this could ever happen. This branch tries to simplify the interaction between the scheduler and its worker threads, reducing the chance for entering an inconsitent state. Maybe the bug is fixed, maybe not.

Includes some extensions that allows to implement [lp1443181](https://bugs.launchpad.net/mixxx/+bug/1443181). I've added a TODO with a reference to this feature request in the code.

uklotzde committed:
Add some assertions about track analysis progress

uklotzde committed:
Delete redundant conditional path

uklotzde committed:
Reset member variables after batch analysis has finished

uklotzde committed:
Wake up worker thread after submitting next track for analysis

uklotzde committed:
Avoid redundant state tracking in analysis scheduler

uklotzde committed:
Keep track of pending tracks in analysis scheduler

This is a precondition for implementing the following feature:
https://bugs.launchpad.net/mixxx/+bug/1443181

uklotzde commented:
Includes some extensions that allows to implement [lp1443181](https://bugs.launchpad.net/mixxx/+bug/1443181). I've added a TODO with a reference to this feature request.

uklotzde committed:
Prepare TrackAnalysisScheduler for future extensions

uklotzde committed:
Prepare implementation of feature request

uklotzde committed:
Fix thread-unsafe invocation of QObject::deleteLater()

uklotzde commented:
Hopefully also the main memory leak is fixed! I suppose it was the thread-unsafe invocation of QObject::deleteLater(). While analyzing 2100 tracks the memory consumption increased from 1.2 g to 1.3 g. GlobalTrackCache constantly stores only 4 tracks to keep the 4 cores busy.

uklotzde committed:
Try to fix CI build on Ubuntu

uklotzde committed:
Merge branch 'master' into unfinished_batch_analysis

uklotzde committed:
Add compatibility workaround for Qt 5.9

uklotzde committed:
Fix and cleanup Qt compatibility includes

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into unfinished_batch_analysis

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into unfinished_batch_analysis

uklotzde commented:
I would like to merge this into master asap for broader testing. Until then I have to classify all upcoming bug reports for master as *obsolete*, because this PR represents the status quo of multi-threaded analysis.

daschuer commented on src/util/compatibility.h:
```diff
@@ -55,4 +55,16 @@ struct QOverload : QConstOverload<Args...>, QNonConstOverload<Args...>
+}
+#else
+inline void invokeDeleteLater(QObject* obj) {
+    // functor
```
A functor is a class with an operator(). This is only a member function pointer.

daschuer commented on src/library/dao/cue.cpp:
```diff
@@ -6,14 +6,19 @@
+//static
+void CuePointer::deleteLater(Cue* pCue) {
+    if (pCue) {
+        // QObject::deleteLater() is not thread-safe!
```
Why do you think that? 
deleteLater() calls QCoreApplication::postEvent() which is thread save.

uklotzde replied:
Because deleteLater() is a regular slot and the documentation does not explicitly state that it is thread-safe to be called from other threads? I didn't check the source code. Just reasoned according to the rules of the [Qt5 documentation](http://doc.qt.io/qt-5/threads-reentrancy.html):

"**Note**: Qt classes are only documented as thread-safe if they are intended to be used by multiple threads. If a function is not marked as thread-safe or reentrant, it should not be used from different threads. If a class is not marked as thread-safe or reentrant then a specific instance of that class should not be accessed from different threads."

daschuer replied:
Ok, so you solution is correct and we can go with it.
However it uses two times the Qt meassage queue.
Did you consider to call QCoreApplication::postEvent() directly?

daschuer reviewed (changes requested):
Thank you.

uklotzde committed:
Handle expected race conditions when receiving queued signals

Events from queued connections may have already implicitly been
handled before, e.g. by subitting tasks to all worker threads that
were ready to receive them at this moment.

uklotzde committed:
Merge branch 'master' into unfinished_batch_analysis

uklotzde committed:
Fix wording

uklotzde committed:
Add a comment about thread-safety of QObject::deleteLater()

uklotzde commented on src/analyzer/analyzerthread.cpp:
```diff
@@ -191,20 +191,28 @@ void AnalyzerThread::doRun() {
+            << nextTrack->getId();
+    if (m_nextTrack.enqueue(std::move(nextTrack))) {
+        // Ensure that the submitted track gets processed
+        // eventually by waking the worker thread up!
```
"eventually" like in "eventually consistent". I will try to revise the comment.

uklotzde committed:
Revise comment about liveliness of worker threads

daschuer commented:
The missing thread savety hint for deleteLater can be IMHO considered as a short coming of the Qt docs. Otherwise the major use case to use it as smart pointer deleter to get arround race condition at the life time end would be sense less. I would prefere to call deleteLater directly and file a QT doc bug instead. What do you think?

uklotzde committed:
Change log level from debug to trace

uklotzde committed:
Delete unnecessary wrapper for QObject::deleteLater()

uklotzde committed:
Merge branch 'master' into unfinished_batch_analysis

uklotzde commented:
Done and waiting for approval.

daschuer commented:
LGTM, Thank you.
